framework/src/kernels/ArrayTimeDerivative.C

```diff
@@ -0,0 +1,84 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ArrayTimeDerivative.h"
+
+registerMooseObject("MooseApp", ArrayTimeDerivative);
+
+template <>
+InputParameters
+validParams<ArrayTimeDerivative>()
+{
+  InputParameters params = validParams<ArrayTimeKernel>();
+  params.addClassDescription("Array time derivative operator with the weak form of $(\\psi_i, "
+                             "\\frac{\\partial u_h}{\\partial t})$.");
+  params.addParam<MaterialPropertyName>("time_derivative_coefficient",
+                                        "The name of the coefficient matrix");
+  MooseEnum opt("scalar=0 array=1 full=2", "array");
+  params.addParam<MooseEnum>(
+      "time_derivative_coefficient_type", opt, "Time derivative coefficient type");
+  return params;
+}
+
+ArrayTimeDerivative::ArrayTimeDerivative(const InputParameters & parameters)
+  : ArrayTimeKernel(parameters),
+    _coeff_type(getParam<MooseEnum>("time_derivative_coefficient_type"))
+{
+  if (_coeff_type == 0)
+    _coeff = &getMaterialProperty<Real>("time_derivative_coefficient");
+  else if (_coeff_type == 1)
```

Comment:
Good idea! I'll switch to using hasMaterialProperty instead of asking for it from the user.

Reply:
Note: I get an error when I move these checks to initialSetup.
`*** ERROR ***`
`Material properties must be retrieved during object construction. This is a code problem.`

Reply:
Okay - this is going to have to go back into the constructor due to quirks of object initialization order.  The constructor only works in this case because the object you are coding is *not* a material.

Reply:
I kept the `getMaterialProperty`s in the constructor but I had to move the `mooseAssert`s into `computeQpResidual` because of an error I was getting in debug mode. It seems I couldn't access the size of `_coeff` until then.

Reply:
That is fine/expected.